services/ingestion/sealing_verifier.go

```diff
@@ -0,0 +1,334 @@
+package ingestion
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/onflow/flow-go-sdk"
+	"github.com/onflow/flow-go-sdk/access"
+	flowGo "github.com/onflow/flow-go/model/flow"
+	"github.com/rs/zerolog"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/onflow/flow-evm-gateway/models"
+	errs "github.com/onflow/flow-evm-gateway/models/errors"
+	"github.com/onflow/flow-evm-gateway/services/requester"
+	"github.com/onflow/flow-evm-gateway/storage/pebble"
+)
+
+var _ models.Engine = &SealingVerifier{}
+
+// SealingVerifier verifies that soft finality events received over the Access polling API match the
+// actually sealed results from the event stream.
+type SealingVerifier struct {
+	*models.EngineStatus
+
+	logger zerolog.Logger
+	client *requester.CrossSporkClient
+	chain  flowGo.ChainID
+
+	startHeight            uint64
+	eventsHash             *pebble.EventsHash
+	unsealedBlocksToVerify map[uint64]flow.Identifier
```

Comment:
Is it possible to have two different event hashes for the same height?

There should be only one block for the same height, because we only process finalized block by height right?

For finalize height, what guarantee we only receive one events hash for the height?

Reply:
the block tracker will visit each block exactly once, so AddBlock should only be called once